### ISSUE TITLE
AttributeError: 'MypyStatusItem' object has no attribute '_obj'

### ISSUE THREAD

**Issue by Nusnus** [dbader/pytest-mypy#113]
Hey,

I'm using both `pytest-mypy` and `allure-pytest` together and when I generate the report, I'm getting weird errors I don't know how to solve.

Can you help me please?

**Test Code**
```python
import allure


def test_mypy1():

    def foo(arg: str) -> int:
        return arg


@allure.severity(allure.severity_level.MINOR)
def test_mypy2():

    def boo() -> int:
        return 10


def test_mypy3():

    def boo() -> int:
        return 'I should have been an int'
```

**Error: AttributeError: 'MypyStatusItem' object has no attribute '_obj'**
```
cls = <class '_pytest.runner.CallInfo'>
func = <function call_runtest_hook.<locals>.<lambda> at 0x7f855c16a3b0>
when = 'setup'
reraise = (<class '_pytest.outcomes.Exit'>, <class 'KeyboardInterrupt'>)

    @classmethod
    def from_call(
        cls,
        func: "Callable[[], TResult]",
        when: "Literal['collect', 'setup', 'call', 'teardown']",
        reraise: Optional[
            Union[Type[BaseException], Tuple[Type[BaseException], ...]]
        ] = None,
    ) -> "CallInfo[TResult]":
        excinfo = None
        start = timing.time()
        precise_start = timing.perf_counter()
        try:
>           result: Optional[TResult] = func()

.venv/lib/python3.7/site-packages/_pytest/runner.py:311: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>       lambda: ihook(item=item, **kwds), when=when, reraise=reraise
    )

.venv/lib/python3.7/site-packages/_pytest/runner.py:255: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <_HookCaller 'pytest_runtest_setup'>, args = ()
kwargs = {'item': <MypyStatusItem mypy>}, notincall = set()

    def __call__(self, *args, **kwargs):
        if args:
            raise TypeError("hook calling supports only keyword arguments")
        assert not self.is_historic()
        if self.spec and self.spec.argnames:
            notincall = (
                set(self.spec.argnames) - set(["__multicall__"]) - set(kwargs.keys())
            )
            if notincall:
                warnings.warn(
                    "Argument(s) {} which are declared in the hookspec "
                    "can not be found in this hook call".format(tuple(notincall)),
                    stacklevel=2,
                )
>       return self._hookexec(self, self.get_hookimpls(), kwargs)

.venv/lib/python3.7/site-packages/pluggy/hooks.py:286: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <_pytest.config.PytestPluginManager object at 0x7f8558758150>
hook = <_HookCaller 'pytest_runtest_setup'>
methods = [<HookImpl plugin_name='nose', plugin=<module '_pytest.nose' from '/Users/nusnus/dev/flask_tester/.venv/lib/python3.7/...855c013450>>, <HookImpl plugin_name='logging-plugin', plugin=<_pytest.logging.LoggingPlugin object at 0x7f855c034b90>>]
kwargs = {'item': <MypyStatusItem mypy>}

    def _hookexec(self, hook, methods, kwargs):
        # called from all hookcaller instances.
        # enable_tracing will set its own wrapping function at self._inner_hookexec
>       return self._inner_hookexec(hook, methods, kwargs)

.venv/lib/python3.7/site-packages/pluggy/manager.py:93: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

hook = <_HookCaller 'pytest_runtest_setup'>
methods = [<HookImpl plugin_name='nose', plugin=<module '_pytest.nose' from '/Users/nusnus/dev/flask_tester/.venv/lib/python3.7/...855c013450>>, <HookImpl plugin_name='logging-plugin', plugin=<_pytest.logging.LoggingPlugin object at 0x7f855c034b90>>]
kwargs = {'item': <MypyStatusItem mypy>}

    self._inner_hookexec = lambda hook, methods, kwargs: hook.multicall(
        methods,
        kwargs,
>       firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
    )

.venv/lib/python3.7/site-packages/pluggy/manager.py:87: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

hook_impls = [<HookImpl plugin_name='nose', plugin=<module '_pytest.nose' from '/Users/nusnus/dev/flask_tester/.venv/lib/python3.7/...855c013450>>, <HookImpl plugin_name='logging-plugin', plugin=<_pytest.logging.LoggingPlugin object at 0x7f855c034b90>>]
caller_kwargs = {'item': <MypyStatusItem mypy>}, firstresult = False

    def _multicall(hook_impls, caller_kwargs, firstresult=False):
        """Execute a call into multiple python functions/methods and return the
        result(s).
    
        ``caller_kwargs`` comes from _HookCaller.__call__().
        """
        __tracebackhide__ = True
        results = []
        excinfo = None
        try:  # run impl and wrapper setup functions in a loop
            teardowns = []
            try:
                for hook_impl in reversed(hook_impls):
                    try:
                        args = [caller_kwargs[argname] for argname in hook_impl.argnames]
                    except KeyError:
                        for argname in hook_impl.argnames:
                            if argname not in caller_kwargs:
                                raise HookCallError(
                                    "hook call must provide argument %r" % (argname,)
                                )
    
                    if hook_impl.hookwrapper:
                        try:
                            gen = hook_impl.function(*args)
                            next(gen)  # first yield
                            teardowns.append(gen)
                        except StopIteration:
                            _raise_wrapfail(gen, "did not yield")
                    else:
                        res = hook_impl.function(*args)
                        if res is not None:
                            results.append(res)
                            if firstresult:  # halt further impl calls
                                break
            except BaseException:
                excinfo = sys.exc_info()
        finally:
            if firstresult:  # first result hooks return a single value
                outcome = _Result(results[0] if results else None, excinfo)
            else:
                outcome = _Result(results, excinfo)
    
            # run all wrapper post-yield blocks
            for gen in reversed(teardowns):
                try:
>                   gen.send(outcome)

.venv/lib/python3.7/site-packages/pluggy/callers.py:203: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <allure_pytest.listener.AllureListener object at 0x7f855c013450>
item = <MypyStatusItem mypy>

    @pytest.hookimpl(hookwrapper=True)
    def pytest_runtest_setup(self, item):
        if not self._cache.get(item.nodeid):
            uuid = self._cache.push(item.nodeid)
            test_result = TestResult(name=item.name, uuid=uuid, start=now(), stop=now())
            self.allure_logger.schedule_test(uuid, test_result)
    
        yield
    
        uuid = self._cache.get(item.nodeid)
        test_result = self.allure_logger.get_test(uuid)
        for fixturedef in _test_fixtures(item):
            group_uuid = self._cache.get(fixturedef)
            if not group_uuid:
                group_uuid = self._cache.push(fixturedef)
                group = TestResultContainer(uuid=group_uuid)
                self.allure_logger.start_group(group_uuid, group)
            self.allure_logger.update_group(group_uuid, children=uuid)
        params = item.callspec.params if hasattr(item, 'callspec') else {}
    
>       test_result.name = allure_name(item, params)

.venv/lib/python3.7/site-packages/allure_pytest/listener.py:88: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

item = <MypyStatusItem mypy>, parameters = {}

    def allure_name(item, parameters):
        name = escape_name(item.name)
>       title = allure_title(item)

.venv/lib/python3.7/site-packages/allure_pytest/utils.py:110: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

item = <MypyStatusItem mypy>

    def allure_title(item):
>       return getattr(item._obj, '__allure_display_name__', None)
E       AttributeError: 'MypyStatusItem' object has no attribute '_obj'

.venv/lib/python3.7/site-packages/allure_pytest/utils.py:34: AttributeError
```

**Comment by dmtucker** [comment]
At first glance, I suspect this is an `allure-pytest` bug.
What versions of `pytest`, `pytest-mypy`, and `allure-pytest` are you using?

**Comment by Nusnus** [comment]
```
allure-pytest==2.8.29
allure-python-commons==2.8.29
pytest==6.2.1
pytest-mypy==0.7.0
```

**Comment by dmtucker** [comment]
Looks like the `_obj` ref was introduced in 2.8.23 (12/2020): https://github.com/allure-framework/allure-python/commit/ed38912d8459102f2f0c032246f4e5f446887c1d#diff-6e193e999f1c57979aef3f32198abb087385243d129deebef6dca1c1a79b56a3R34

**Comment by dmtucker** [comment]
`obj` ([which should probably be used instead of `_obj` directly...](https://github.com/pytest-dev/pytest/blob/5336ba28ccc583f4cbde0082ee1d0bcc4e2002eb/src/_pytest/python.py#L291-L305)) is defined on a traditionally-defined pytest test function, but not on a pytest-mypy-generated Item:
```
(Pdb) pp items
[<MypyFileItem mypy0.782>, <Function test_fail>, <Function test_pass>]
(Pdb) pp items[0].obj
*** AttributeError: 'MypyFileItem' object has no attribute 'obj'
(Pdb) pp items[1].obj
<function test_fail at 0x7ff2eb3cb790>
(Pdb) pp items[1].obj == items[1]._obj
True
```

**Comment by dmtucker** [comment]
Considering `.obj` returns the function object for a pytest test and there is no function object to return for a `MypyItem`, I don't think the issue is on the pytest-mypy side. In the allure-pytest code, the failing function, `allure_title` may return `None`. Perhaps a patch to make that happen if `.obj` is not defined is in order?

**Comment by dmtucker** [comment]
This has been fixed in allure-pytest-2.8.31: https://github.com/allure-framework/allure-python/blob/163f08c1c9009f87491e1767243169ad857eaa92/allure-pytest/src/utils.py#L33-L37

@Nusnus could you give that version a go and confirm that things work correctly now?

**Comment by Nusnus** [comment]
Yes it does! @dmtucker 
Thank you very much! Very fast response. Much appreciated.

**Details**
`platform linux -- Python 3.7.9, pytest-6.2.1, py-1.10.0, pluggy-0.13.1 -- /usr/bin/python3`
`plugins: mypy-0.8.0, allure-pytest-2.8.31`

`test_mypyallurebug.py`:
```python
import allure


def test_mypy1():

    def foo(arg: str) -> int:
        return arg


@allure.severity(allure.severity_level.MINOR)
def test_mypy2():

    def boo() -> int:
        return 10


def test_mypy3():

    def boo() -> int:
        return 'I should have been an int'
```
**Results**
```
pytest_mypy.MypyError:
7: error: Incompatible return value type (got "str", expected "int")
20: error: Incompatible return value type (got "str", expected "int")
```
